src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java

```diff
@@ -63,70 +70,151 @@ public void dispose() {
+                            if (serviceReference.equals(reference.getServiceReference())) {
+                                ServiceType service = (ServiceType) reference.getService();
+                                if (service != null) {
+                                    result.add(service);
+                                }
+                                break;
+                            }
+                        }
+                    } else {
+                        // concurrent change; restart
+                        return getServices(type, filter);
+                    }
                 }
             }
-
-            // unget the service if another thread was faster
-            if (ref != null) {
-                this.bundleContext.ungetService(ref);
-            }
+        } catch (InvalidSyntaxException e) {
+            LOGGER.error(String.format("Unable to retrieve the services of type %s.", type.getName()), e);
         }
-
-        // return whatever we got (which may be null)
-        return (ServiceType) reference.service;
+        if (!result.isEmpty()) {
+            ServiceType[] srv = (ServiceType[]) Array.newInstance(type, result.size());
+            return result.toArray(srv);
+        }
+        return null;
     }
 
     /**
      * @see org.osgi.framework.ServiceListener#serviceChanged(org.osgi.framework.ServiceEvent)
      */
     public void serviceChanged(ServiceEvent event) {
-        final String[] objectClasses = (String[])event.getServiceReference().getProperty(Constants.OBJECTCLASS);
-        if ( objectClasses != null) {
-            for(final String key : objectClasses) {
-                Reference ref = null;
-                synchronized ( this ) {
-                    ref = this.cache.remove(key);
+        ServiceReference<?> serviceReference = event.getServiceReference();
+        final String[] objectClasses = (String[]) serviceReference.getProperty(Constants.OBJECTCLASS);
+        if (objectClasses != null) {
+            for (final String key : objectClasses) {
+                SortedSet<Reference> references;
+                synchronized (this) {
+                    references = this.cache.remove(key);
+                    if (references != null) {
+                        for (Reference reference : references) {
+                            bundleContext.ungetService(reference.getServiceReference());
+                        }
+                    }
                 }
-                if ( ref != null && ref != NULL_REFERENCE ) {
-                    this.bundleContext.ungetService(ref.reference);
+            }
+        }
+    }
+
+    private <ServiceType> SortedSet<Reference> getCachedReferences(Class<ServiceType> type) {
+        String key = type.getName();
+        SortedSet<Reference> references = cache.get(key);
+        if (references == null) {
+            references = Collections.synchronizedSortedSet(new TreeSet<>((o1, o2) -> -1 * o1.compareTo(o2)));
```

Comment:
Doubt it matters much but should we maybe use a ConcurrentSkipListSet?

Reply:
For the comparator, you could use [`Comparator.reverseOrder()`](https://docs.oracle.com/javase/8/docs/api/java/util/Comparator.html#reverseOrder--) instead. IMHO it's more readable.

Reply:
Got rid of this in [712b011](https://github.com/apache/sling-org-apache-sling-scripting-core/pull/13/commits/712b0115040cf239d59eeb8baad13a175ff17309).